docs/design/federated-replicasets.md

```diff
@@ -0,0 +1,542 @@
+      [[I6]](#heading=h.z90979gc2216).
+
++ [I3] For each of the cluster start an store and informer to monitor
+   whether the created pods are eventually scheduled and what is the
+   current number of correctly running ready pods. Errors:
+
+   + [E7] It is impossible to tell that an informer lost connection
+      with a remote cluster or has other synchronization problem so it
+      should be handled by cluster liveness probe and deletion
+      [[I6]](#heading=h.z90979gc2216)
+
++ [I4] It is assumed that a not scheduled pod is a normal situation
+and can last up to X min if there is a huge traffic on the
+cluster. However if the replicas are not scheduled in that time then
+FRSC should consider moving most of the unscheduled replicas
+elsewhere. For that purpose FRSC will maintain a data structure
+where for each FRS controlled LRS we store a list of pods belonging
+to that LRS along with their current status and status change timestamp.
+
++ [I5] If a new cluster is added to the federation then it doesn’t
+   have a LRS and the situation is equal to
+   [[E1]](#heading=h.wn3dfsyc4yuh)/[[E4]](#heading=h.vlyovyh7eef).
+
++ [I6] If a cluster is removed from the federation then the situation
+   is equal to multiple [E4]. It is assumed that if a connection with
+   a cluster is lost completely then the cluster is removed from the
+   the cluster list (or marked accordingly) so
+   [[E6]](#heading=h.in6ove1c1s8f) and [[E7]](#heading=h.37bnbvwjxeda)
+   don’t need to be handled.
+
++ [I7] All ToBeChecked FRS are browsed every 1 min (configurable),
+   checked against the current list of clusters, and all missing LRS
+   are created. This will be executed in combination with [I8].
+
++ [I8] All pods from ToBeChecked FRS/LRS are browsed every 1 min
+   (configurable) to check whether some replica move between clusters
+   is needed or not.
+
++  FRSC never moves replicas to LRS that have not scheduled/running
+pods or that has pods that failed to be created.
+
+   + When FRSC notices that a number of pods are not scheduler/running
+      or not_even_created in one LRS for more than Y minutes it takes
+      most of them from LRS, leaving couple still waiting so that once
+      they are scheduled FRSC will know that it is ok to put some more
+      replicas to that cluster.
+
++   [I9] FRS becomes ToBeChecked if:
+   +  It is newly created
+   +  Some replica set inside changed its status
+   +  Some pods inside cluster changed their status
+   +  Some cluster is added or deleted.
+> FRS stops ToBeChecked if is in desired configuration (or is stable enough).
+
+## (RE)Scheduling algorithm
+
+To calculate the (re)scheduling moves for a given FRS:
+
+1. For each cluster FRSC calculates the number of replicas that are placed 
+(not necessary up and running) in the cluster and the number of replicas that 
+failed to be scheduled. Cluster capacity is the difference between the
+the placed and failed to be scheduled.
```

Comment:
Sorry, I don't understand this sentence.  Perhaps I'm being dumb.  How can cluster capacity be defined this way? e.g. If I try to schedule 10, and 9 schedule, leaving 1 unscheduled, it seems that the remaining capacity should be zero, not 9-1=8.  Does this sentence need rewording?